### PR TITLE
Client sided gun code

### PR DESCRIPTION
**Pull request by Luke0x22918**

### CODE REVIEW

File: src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua [hunk 1]
```diff
@@ -1,9 +1,15 @@
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+
+local Gun = require(ReplicatedStorage.Client.Tool.Gun)
+
 local AssaultRifle = {}
 AssaultRifle.__index = AssaultRifle
+setmetatable(AssaultRifle, Gun)
+AssaultRifle.Config = require(ReplicatedStorage.Common.Config.Guns.AssaultRifle)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua]: Import this explicitly in the requires as 'AssultRifleConfig'

File: src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua [hunk 2]
```diff
@@ -1,9 +1,15 @@
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+
+local Gun = require(ReplicatedStorage.Client.Tool.Gun)
+
 local AssaultRifle = {}
 AssaultRifle.__index = AssaultRifle
+setmetatable(AssaultRifle, Gun)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua]: Combine these to local AssaultRifle = setmetatable({}, Gun) to follow the inheritance constructor idiom

File: src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
+AssaultRifle.Config = require(ReplicatedStorage.Common.Config.Guns.AssaultRifle)
 
-function AssaultRifle.new(instance)
-	local self = setmetatable({}, AssaultRifle)
-	self.Instance = instance
+function AssaultRifle.new(instance: Tool)
+	local self = Gun.new(instance)
+	setmetatable(self, AssaultRifle)
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Tools/AssaultRifle.lua]: Combine these to local self = setmetatable(Gun.new(instance), AssultRifle) to follow the inheritance constructor idiom

File: src/ReplicatedStorage/Client/InstanceComponents/Tools/SMG.lua [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
+SMG.Config = require(ReplicatedStorage.Common.Config.Guns.SMG)
 
-function SMG.new(instance)
-	local self = setmetatable({}, SMG)
-	self.Instance = instance
+function SMG.new(instance: Tool)
+	local self = Gun.new(instance)
+	setmetatable(self, SMG)
 	return self
 end
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Tools/SMG.lua]: This has all of the same issues as AssaultRifle

File: src/ReplicatedStorage/Client/InstanceComponents/Tools/Shotgun.lua [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
+setmetatable(Shotgun, Gun)
+Shotgun.Config = require(ReplicatedStorage.Common.Config.Guns.Shotgun)
 
 function Shotgun.new(instance)
-	local self = setmetatable({}, Shotgun)
-	self.Instance = instance
+	local self = Gun.new(instance)
+	setmetatable(self, Shotgun)
 	return self
 end
```
Comment (4812571) [inline src/ReplicatedStorage/Client/InstanceComponents/Tools/Shotgun.lua]: This has all of the same issues as AssaultRifle

File: src/ReplicatedStorage/Common/Config/Guns/AssaultRifle.lua [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+	ScopeFOV = 60,
+	BulletDecoration = {
+		Color = Color3.fromRGB(245, 205, 48),
+		Thickness = 0.15
+	},
+	FireSound = {
+		SoundId = "rbxassetid://131070686",
+		Volume = 0.5
+	}
+}
```
Comment (4812571) [inline src/ReplicatedStorage/Common/Config/Guns/AssaultRifle.lua]: Explicitly declare this as local AssaultRifleConfig = {

}

return AssaultRifleConfig

-- see other configs to see the pattern :)

File: src/ReplicatedStorage/Common/Config/Guns/SMG.lua [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	BulletDecoration = {
+		Color = Color3.fromRGB(245, 205, 48),
+		Thickness = 0.15,
+		ImpulseForce = 5
+	},
+	FireSound = {
+		SoundId = "rbxassetid://168143115",
+		Volume = 0.3
+	}
+}
```
Comment (4812571) [inline src/ReplicatedStorage/Common/Config/Guns/SMG.lua]: This has the same issues as AssaultRifle Config

File: src/ReplicatedStorage/Common/Config/Guns/Shotgun.lua [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	BulletDecoration = {
+		Color = Color3.fromRGB(245, 205, 48),
+		Thickness = 0.15,
+		ImpulseForce = 30
+	},
+	FireSound = {
+		SoundId = "rbxassetid://181135016",
+		Volume = 0.5
+	}
+}
```
Comment (4812571) [inline src/ReplicatedStorage/Common/Config/Guns/Shotgun.lua]: This has the same issues as AssaultRifle Config

File: src/ServerScriptService/Services/BulletHandlerService/init.lua [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        if player.Character then
+            task.spawn(characterAdded, player.Character)
+        end
+    end
+
+    Players.PlayerAdded:Connect(playerAdded)
+
+    for _, player in Players:GetPlayers() do
+        task.spawn(playerAdded, player)
+    end
```
Comment (4812571) [inline src/ServerScriptService/Services/BulletHandlerService/init.lua:42]: This is crazy - but temporary! I'll allow it. -- Reminder to open an issue.

File: src/ServerScriptService/Services/BulletHandlerService/init.lua [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+end
+
+function BulletHandlerService:KnitInit()
+    TempCodeToGiveGunsRemoveLaterOkayThxBye()
```
Comment (4812571) [inline src/ServerScriptService/Services/BulletHandlerService/init.lua]: This should be in :KnitStart()!!!

File: src/ReplicatedStorage/Client/Tool/init.lua [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	self.Instance.Unequipped:Connect(function(...)
+		mouse.Icon = ""
+		self:OnUnequipped(...)
+	end)
+end
+
+-- public:
+
+function Tool:OnActivated()
+	-- virtual method
```
Comment (4812571) [inline src/ReplicatedStorage/Client/Tool/init.lua]: Could we maybe use the comment '-- Virtual' as a standard? These are our first virtual methods in this codebase and I want to get a standard for it.

Comment (4812571) [reply src/ReplicatedStorage/Client/Tool/init.lua]: (as opposed to -- virtual method)

**Review by 4812571 — CHANGES_REQUESTED** [review]
Some formatting changes, then we're good to go!

File: src/ServerScriptService/Services/BulletHandlerService/init.lua (hunk 10 above)
Comment (Hexcede) [inline src/ServerScriptService/Services/BulletHandlerService/init.lua]: Later this should really go into the `Character` component.

**Review by Luke0x22918 — COMMENTED** [review]
Fixed formatting

File: src/ServerScriptService/Services/BulletHandlerService/init.lua (hunk 10 above)
Comment (4812571) [reply src/ServerScriptService/Services/BulletHandlerService/init.lua]: Semi factual

Comment (4812571) [reply src/ServerScriptService/Services/BulletHandlerService/init.lua]: The calls themselves should, but the implementation will probably go into a service

**Review by 4812571 — CHANGES_REQUESTED** [review]
Just those last few changes :)

**Review by 4812571 — APPROVED** [review]
LGTM!

**Pull request merged**